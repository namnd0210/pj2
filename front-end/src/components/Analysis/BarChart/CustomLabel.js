import React from 'react'

export default function CustomLabel(data) {
  const index = data.index;
  const xValue = data.data[index].x;
  const yValue = data.data[index].y;
  const x = data.x;
  const y = data.y;
  const isLeft = data.isLeft;
  return (
    <svg>
      {
        isLeft && <text x={x + 12} y={y - 15} style={{ fontSize: "25" }}>
          {xValue}
        </text>
      }
      {
        !isLeft && <text x={x + 610} y={y - 15} style={{ direction: "rtl", fontSize: "25" }}>
          {yValue}
        </text>
      }
    </svg>
  )
}
