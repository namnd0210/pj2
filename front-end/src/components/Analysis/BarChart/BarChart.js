import React, { useEffect } from 'react';
import {
  VictoryChart, VictoryBar, VictoryAxis, VictoryLabel
} from 'victory';
import _ from 'lodash';
import { useDispatch, useSelector } from "react-redux";

import { fetchBarChartData } from '../../../actions';
import CustomLabel from './CustomLabel';
import Loading from '../../Loading';

export default function Ranking({ startDate, endDate }) {
  const { data, isLoading } = useSelector(state => state.barChartData)
  const dispatch = useDispatch()

  useEffect(() => {
    dispatch(fetchBarChartData(startDate, endDate));
  }, [startDate, endDate])

  const handledData = _.sortBy(data, ['y'], ['desc']);

  return (
    <div>
      <h2>Ranking</h2>
      {isLoading && <Loading color="#eb09eb" />}
      {!isLoading &&
        <svg viewBox="0 0 700 500">
          <text x={51} y={25} style={{ fontSize: "25" }}>Day</text>
          <text x={560} y={25} style={{ fontSize: "25" }}>Number</text>
          <g transform={'translate(0, 50)'}>
            <VictoryChart width={700} height={450}
              standalone={false}
            >

              <VictoryBar horizontal
                alignment="start"
                style={{ data: { fill: '#ededed' } }}
                standalone={false}
                data={handledData}
                barRatio={1.5}
              />
              <VictoryAxis dependentAxis
                standalone={false}
                style={{
                  axis: { stroke: "#fff" },
                  tickLabels: { fill: "none" },
                }}
              />
              <VictoryAxis
                standalone={false}
                style={{
                  axis: { stroke: "#fff" },
                  grid: { stroke: "#c9cdd4" },
                  tickLabels: { fill: "none" }
                }}
                tickLabelComponent={
                  <CustomLabel
                    data={handledData}
                    isLeft={true}
                  />}
              />
              <VictoryAxis
                standalone={false}
                style={{
                  axis: { stroke: "#fff" }
                }}
                tickLabelComponent={
                  <CustomLabel
                    data={handledData}
                    isLeft={false}
                  />}
              />
            </VictoryChart>
          </g>
        </svg>}
    </div>
  )
}