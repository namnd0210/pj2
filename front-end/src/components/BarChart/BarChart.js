import React, { useEffect } from 'react';
import {
  VictoryChart, VictoryBar, VictoryAxis, VictoryLabel
} from 'victory';
import _ from 'lodash';
import { useDispatch, useSelector } from "react-redux";

import { fetchBarChartData } from '../../actions';
import CustomLabel from './CustomLabel';
import Loading from '../Loading';

export default function Ranking({ startDate, endDate }) {
  const { data, isLoading } = useSelector(state => state.barChartData)
  const dispatch = useDispatch()

  useEffect(() => {
    dispatch(fetchBarChartData(startDate, endDate));
  }, [startDate, endDate])

  const handledData = _.sortBy(data, ['y'], ['desc']);

  return (
    <div>
      <h2>Ranking</h2>
      {isLoading && <Loading color="#eb09eb" />}
      {!isLoading &&
        <svg viewBox="0 -30 500 500">
          <VictoryChart
            standalone={false}
          >
            <VictoryLabel x={51} y={0}
              standalone={false}
              text="Day"
            />
            <VictoryLabel x={340} y={0}
              standalone={false}
              text="Number"
            />
            <VictoryBar horizontal
              alignment="start"
              style={{ data: { fill: '#ededed' } }}
              standalone={false}
              data={handledData}
              barRatio={1.5}
            />
            <VictoryAxis dependentAxis
              standalone={false}
              style={{
                axis: { stroke: "#fff" },
                tickLabels: { fill: "none" },
              }}
            />
            <VictoryAxis
              standalone={false}
              style={{
                axis: { stroke: "#fff" },
                grid: { stroke: "#c9cdd4" },
                tickLabels: { fill: "none" }
              }}
              tickLabelComponent={
                <CustomLabel
                  data={handledData}
                  isLeft={true}
                />}
            />
            <VictoryAxis
              standalone={false}
              style={{
                axis: { stroke: "#fff" }
              }}
              tickLabelComponent={
                <CustomLabel
                  data={handledData}
                  isLeft={false}
                />}
            />
          </VictoryChart>}
      </svg>}
    </div>
  )
}