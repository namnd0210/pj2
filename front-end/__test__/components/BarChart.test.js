import React from 'react'
import Enzyme, { shallow } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import BarChart from '../../src/components/Analysis/BarChart/BarChart'

Enzyme.configure({ adapter: new Adapter() })

jest.mock('react-redux', () => ({
  useDispatch: () => { },
  useSelector: () => ({})
}));

describe('<BarChart />', () => {
  it('Expect to not log errors in console', () => {
    const spy = jest.spyOn(global.console, 'error');
    const wrapper = shallow(
      <BarChart />,
    );
    expect(wrapper).not.toBeNull();
    expect(spy).not.toHaveBeenCalled();
  })
  it('Should render and match the snapshot', () => {
    const wrapper = shallow(
      <BarChart />,
    );
    expect(wrapper).toMatchSnapshot();
  })
})