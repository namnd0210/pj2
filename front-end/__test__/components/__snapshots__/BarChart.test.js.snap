// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BarChart /> Should render and match the snapshot 1`] = `ShallowWrapper {}`;
